### PR TITLE
Bypass shape inference in deconv2d and use the output shape provided by the user

### PR DESCRIPTION
**Pull request by tboquet**
Following the discussions in #3540 and #3824 this pull request removes the shape inference part of the `Deconv2d` layer and use the output shape provided by the user.

Because the user must specify an output shape and the shape inference introduced does not work in all cases, the code is more concise and works in cases where the user wants to introduce operations where a precise output shape is needed.

@yaringal is this reasonnable?

### CODE REVIEW

**Comment by EderSantana** [comment]
quick experiments request, can you make the variational conv example deeper? I mean two conv dows and two convs up, hourglass style. That used to file under theano for me

**Comment by tboquet** [comment]
Sure, I wrote something that is working on Tensorflow and I'll test it on Theano today. Not sure on the architecture though, the results both for the existing example and the one with the modification I made lead to poor results.

**Comment by yaringal** [comment]
Apologies for not responding - I'm currently writing up and won't have time to help until the end of the month... 

I answered the discussion here https://github.com/fchollet/keras/issues/3824. In short, I would say that atm the shape inference is not that useful and it might make sense to remove it (but say explicitly in the doc string that the output shape is not arbitrary and should be inferred from the input e.g. using a dummy variable).

**Comment by tboquet** [comment]
@yaringal thanks! I will add a note in the doc string and maybe an example on how to manually check the output shape.

**Comment by fchollet** [comment]
I'll merge it when you consider it to be ready.

File: examples/variational_autoencoder_deconv.py [hunk 1]
```diff
@@ -15,21 +15,30 @@
 # input image dimensions
 img_rows, img_cols, img_chns = 28, 28, 1
 # number of convolutional filters to use
-nb_filters = 32
+nb_filters = 64
 # convolution kernel size
 nb_conv = 3
 
-batch_size = 16
+batch_size = 128
```
Comment (tboquet) [inline examples/variational_autoencoder_deconv.py]: I made the batch size bigger so the training is faster.

File: examples/variational_autoencoder_deconv.py [hunk 2]
```diff
@@ -15,21 +15,30 @@
...
 original_dim = (img_chns, img_rows, img_cols)
 latent_dim = 2
 intermediate_dim = 128
-epsilon_std = 0.01
+epsilon_std = 1.0
```
Comment (tboquet) [inline examples/variational_autoencoder_deconv.py]: To be able to compare the results with the other variationnal autoencoder example I also changed the std.

File: examples/variational_autoencoder_deconv.py [hunk 3]
```diff
@@ -48,23 +57,35 @@ def sampling(args):
...
+c_3_decoded = decoder_c_3(c_2_decoded)
+x_decoded_mean = decoder_mean(c_3_decoded)
+x_decoded_mean = decoder_mean_c(x_decoded_mean)
 
 def vae_loss(x, x_decoded_mean):
     # NOTE: binary_crossentropy expects a batch_size by dim for x and x_decoded_mean, so we MUST flatten these!
     x = K.flatten(x)
     x_decoded_mean = K.flatten(x_decoded_mean)
-    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
+    xent_loss = 28 * 28 * objectives.binary_crossentropy(x, x_decoded_mean)
```
Comment (tboquet) [inline examples/variational_autoencoder_deconv.py]: Same loss as in variationnal_autoencoder.py

**Comment by tboquet** [comment]
It seems there is less noise in the new version of the example.

Old example:
before
New example:
after

@EderSantana, what do you think?

Another thing, why do we want to go as far as 15 stds?

**Comment by tboquet** [comment]
@lukovkin  would you take a look at what I wrote in the doc? Not sure if it's enough.

**Comment by jocicmarko** [comment]
Shouldn't we omit the first element of output_shape tuple? It's always batch size which is the same across all layers.
e.g. ```output_shape=(None, 3, 14, 14)``` -> ```output_shape=(3, 14, 14)```

File: examples/variational_autoencoder_deconv.py (hunk 3 above)
Comment (EderSantana) [inline examples/variational_autoencoder_deconv.py]: isn't this the same as increasing the learning rate? or at least, increasing the importance of the reconstruction error over kl-divergence regularizer. I didn't remember this was the original cost.

**Comment by EderSantana** [comment]
I think is important to pass the batch and everything. Tensorflow doesn't handle reshaping well if there is a None dimension. Plus, if we allow the batch in the output shape we cover the case of deconv in the first layer (if you are splitting the encoder and decoder as different keras models for ex)

File: examples/variational_autoencoder_deconv.py (hunk 3 above)
Comment (tboquet) [inline examples/variational_autoencoder_deconv.py]: You are right it wasn't there before and it seems this weight is necessary for the model to perform correctly. I wasn't sure about this but I tested with and without and I get results similar to  #3373 if I don't use it. I'm not sure about the magnitude of the weight though. I tested with other weights and it made the example work properly too. It was added in [the other example](https://github.com/fchollet/keras/blob/master/examples/variational_autoencoder.py#L42) with #3383. It's related to the `epsilon_std`. I can change this hyperparameter in both examples so that the loss is untouched and the only hyperparameter to tune is the standard deviation of the epsilon. What do you think?

Comment (EderSantana) [inline examples/variational_autoencoder_deconv.py]: I haven't used this scaling factor before. If the papers said it is the appropriate format, we should use it tho

Comment (tboquet) [inline examples/variational_autoencoder_deconv.py]: Right, it's better to use the same settings as in the paper and to remove this arbitrary importance weight. From the paper:
"All parameters, both variational and generative, were initialized by random sampling from N (0, 0.01)".

**Comment by jocicmarko** [comment]
@EderSantana thanks for the input Eder, that makes sense

**Comment by tboquet** [comment]
@fchollet, the PR is ready if you think the doc is clear enough.

For the examples, the results look good:

old `variationnal_autoencodeur_deconv.py` example:
old

new `variationnal_autoencodeur_deconv.py` example:
new

old `variationnal_autoencodeur.py` example:
old

new`variationnal_autoencodeur.py` example:
new

File: examples/variational_autoencoder.py [hunk 4]
```diff
@@ -39,7 +41,7 @@ def sampling(args):
 
 
 def vae_loss(x, x_decoded_mean):
-    xent_loss = original_dim * objectives.binary_crossentropy(x, x_decoded_mean)
+    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
```
Comment (lukovkin) [inline examples/variational_autoencoder.py]: Are we sure about this removal of the multiplication by the original_dim?
@RuiShu added this multiplication in #3220, and he had some arguments to do it.
Initial implementation was without this multiplication.

Comment (tboquet) [inline examples/variational_autoencoder.py]: You are right, I just read the conversation and the paper again. I was a little bit confused about this multiplication because it's not in the example with deconvolution in the master branch. Let me reintroduce this weight!

**Comment by tboquet** [comment]
@lukovkin I changed the `vae_loss` in both examples and replaced the dim ordering option in the doc.

File: examples/variational_autoencoder.py (hunk 4 above)
Comment (EderSantana) [inline examples/variational_autoencoder.py]: BTW, the autoencoding variational bayes paper doesn't cover convolutions, does it? Shouldn't we also add a ref to the paper that first does conv+vae?

Comment (EderSantana) [inline examples/variational_autoencoder.py]: In any case, this PR looks good to me!

Comment (tboquet) [inline examples/variational_autoencoder.py]: Thanks for the review @EderSantana! Do you know any reference using conv+vae? I found [this one](http://people.ee.duke.edu/~lcarin/Yunchen_nips_2016.pdf). The encoder-decoder part seems similar to the example.

Comment (RuiShu) [inline examples/variational_autoencoder.py]: For papers that are closer to the heart of VAEs: 
Might be useful to cite Kingma's [IAF paper](https://arxiv.org/abs/1606.04934).
Larsen's [vae-gan paper](https://arxiv.org/abs/1512.09300) is good as well.

Comment (EderSantana) [inline examples/variational_autoencoder.py]: IAF was an earlier reference. it uses conditional embedding but it is still fine I think. vae-gan is beyond the scope

Comment (tboquet) [inline examples/variational_autoencoder.py]: Right, the first paper is something else, so I will put the other Kingma's paper. Thanks guys!

Comment (yaringal) [inline examples/variational_autoencoder.py]: I can't see the relation to Normalising Flows (IAF paper) or VAE-GANs. If you're looking for VAE papers that make use of convolutional architectures then there are dozens of papers. I'm not sure if the ref needs to be changed as the adaptation is minor (using convolutions for image data instead of inner product layers).

Comment (tboquet) [inline examples/variational_autoencoder.py]: @yaringal , true that it's not a big change. The main focus of this PR is on the change in the shape inference part and it's not a big deal to keep the same reference. If a paper is of particular interest it could maybe be integrated later. I'll let you guys decide!

Comment (EderSantana) [inline examples/variational_autoencoder.py]: @yaringal I thought IAF was using KL in the hidden space and reparametrization trick. If that is not true than we could skip that reference.

**Comment by tboquet** [comment]
@fchollet I think the PR is ready to be merged, a reference could be added later for the exemple.

File: examples/variational_autoencoder_deconv.py [hunk 5]
```diff
@@ -48,23 +57,38 @@ def sampling(args):
 
 # we instantiate these layers separately so as to reuse them later
 decoder_h = Dense(intermediate_dim, activation='relu')
-decoder_f = Dense(nb_filters*img_rows*img_cols, activation='relu')
-decoder_c = Reshape((nb_filters, img_rows, img_cols))
-decoder_mean = Deconvolution2D(img_chns, nb_conv, nb_conv,
-                               (batch_size, img_chns, img_rows, img_cols),
-                               border_mode='same')
+decoder_f = Dense(nb_filters*14*14, activation='relu')
```
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: Spaces around operators

File: examples/variational_autoencoder_deconv.py [hunk 6]
```diff
@@ -48,23 +57,38 @@ def sampling(args):
...
+                              (batch_size, nb_filters, 14, 14),
+                              border_mode='same',
+                              subsample=(1, 1),
+                              activation='relu')
+decoder_mean = Deconvolution2D(nb_filters, 2, 2,
+                               (batch_size, nb_filters, 29, 29),
+                               border_mode='valid',
+                               subsample=(2, 2),
+                               activation='relu')
+decoder_mean_c = Convolution2D(img_chns, 2, 2, border_mode='valid', activation='sigmoid')
```
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: I think these layers would benefit from having more transparent names.

File: examples/variational_autoencoder_deconv.py (hunk 3 above)
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: This should be `img_rows * img_cols`

File: examples/variational_autoencoder_deconv.py [hunk 7]
```diff
@@ -78,6 +102,17 @@ def vae_loss(x, x_decoded_mean):
...
 x_test = x_test.astype('float32')[:, None, :, :] / 255.
 
+end = len(x_test) - len(x_test) % batch_size
+x_test = x_test[:end]
+
+end = len(x_train) - len(x_train) % batch_size
+x_train = x_train[:end]
+
+end = len(y_test) - len(y_test) % batch_size
+y_test = y_test[:end]
```
Comment (fchollet) [inline examples/variational_autoencoder_deconv.py]: Maybe better to simply pick an appropriate batch size?

File: examples/variational_autoencoder.py (hunk 4 above)
Comment (RuiShu) [inline examples/variational_autoencoder.py]: @EderSantana Perhaps I'm misunderstanding, by my impression is that all of the VAE papers use KL + reparameterization trick. It just so happens that some reparameterization tricks (i.e. diagonal gaussian reparam) are easier than others (e.g. normalizing flow).

**Comment by tboquet** [comment]
@fchollet thank you for the review, I think I addressed all the comments, is it good and do you see other things to improve?

**Comment by fchollet** [comment]
LGTM, thanks
